SingleCellPatch/extract_patches.py

```diff
@@ -322,51 +209,103 @@ def process_site_extract_patches(site_path,
             remove_mask, tm, tm2 = generate_mask(positions, 
                                                  positions_labels, 
                                                  cell_id, 
                                                  window, 
                                                  window_segmentation)
             # Select submatrix from the whole slice
-            output_mat = select_window(raw_image, window, padding=0)
+            # output_mat = select_window(raw_image, window, padding=0)
+            output_mat = select_window(raw_image, window, padding=None)
             masked_output_mat = output_mat * (1 - remove_mask) + background_filling * remove_mask
             # To prevent backward compatibility issue cast to int64 and float 64 respectively
             # TODO: solve compatibility issue here
             output_mat = np.concatenate([output_mat, tm, tm2], 2).astype('int64')
             masked_output_mat = np.concatenate([masked_output_mat, tm, tm2], 2).astype('float64')
             site_data[cell_name] = {"mat": output_mat, "masked_mat": masked_output_mat}
+            if save_fig:
+                im_phase = output_mat[:, :, 0]
+                im_phase_masked = masked_output_mat[:, :, 0]
+                # replace zero-padding with min for display
+                im_phase[im_phase == 0] = np.nanmin(im_phase[im_phase != 0])
+                im_phase_masked[im_phase_masked == 0] = np.nanmin(im_phase_masked[im_phase_masked != 0])
+                im_phase = im_adjust(im_phase)
+                im_phase_masked = im_adjust(im_phase_masked)
+                n_rows = 2
+                n_cols = 2
+                fig, ax = plt.subplots(n_rows, n_cols, squeeze=False)
+                ax = ax.flatten()
+                fig.set_size_inches((15, 5 * n_rows))
+                axis_count = 0
+                for im, name in zip([im_phase, im_phase_masked, tm, tm2],
+                                    ['output_mat', 'masked_output_mat', 'tm', 'tm2']):
+                    ax[axis_count].imshow(np.squeeze(im), cmap='gray')
+                    ax[axis_count].axis('off')
+                    ax[axis_count].set_title(name, fontsize=12)
+                    axis_count += 1
+                fig.savefig(os.path.join(site_supp_files_folder, 'patch_t%d_id%d.jpg' % (t_point, cell_id)),
+                            dpi=300, bbox_inches='tight')
+                plt.close(fig)
+        # remove cells that don't have patches, update cell_positions
+        cell_positions_t = []
+        for cells in [mg_cells, non_mg_cells, other_cells]:
+            cells = [cell for cell in cells if cell[0] in cells_to_keep]
+            cell_positions_t.append(cells)
+        cell_positions[t_point] = cell_positions_t
         with open(os.path.join(site_supp_files_folder, 'stacks_%d.pkl' % t_point), 'wb') as f:
             pickle.dump(site_data, f)
+        with open(os.path.join(site_supp_files_folder, 'cell_positions.pkl'), 'wb') as f:
```

Comment:
I actually suggest not overwriting the intermediate pickles, which could make backtrack a bit complicated. We could either use another name or skip this step and do the check (of whether a cell patch is saved) in later steps when needed.

Reply:
`cell_positions` need to be updated to make the downstream processing work because not every patch is saved now with `skip_boundary=True`. We could save `cells_to_keep` and use it to index `cell_positions` later but I'm leaning towards to reduce the number of intermediate pickles/variable if possible. Do you need the original `cell_positions` for any downstream processing?

Reply:
OK in that case it would be fine to keep the current form. I think as long as we keep everything consistent it should be fine. One more issue is that the original `cell_positions.pkl` generated by instance clustering is separated into 3 categories: mg, non-mg and intermediate. As the current pipeline only look at cell identities during analysis, it would be wise just to save everything together. 

It might be easier for me to make these changes, would you mind I handling all these changes in the scripts under `SingleCellPatch`? I will do a quick PR today.

Reply:
@smguo PR #18 fixed these scripts, please note if there is any conflict.